geonode/thumbs/utils.py

```diff
@@ -42,3 +54,124 @@ def assign_missing_thumbnail(instance: Union[Layer, Map]) -> None:
+    styles: str = None,
+    width: int = 240,
+    height: int = 200,
+) -> str:
+    """
+    Method constructing a GetMap URL to the OGC server.
+
+    :param ogc_server_location: OGC server URL
+    :param layers: layers which should be fetched from the OGC server
+    :param bbox: area's bounding box in format: [west, east, south, north, CRS]
+    :param wms_version: WMS version of the query
+    :param mime_type: mime type of the returned image
+    :param styles: styles, which OGC server should use for rendering an image
+    :param width: width of the returned image
+    :param height: height of the returned image
+
+    :return: GetMap URL
+    """
+    # create GetMap query parameters
+    params = {
+        "service": "WMS",
+        "version": wms_version,
+        "request": "GetMap",
+        "layers": ",".join(layers),
+        "bbox": ",".join([str(bbox[0]), str(bbox[2]), str(bbox[1]), str(bbox[3])]),
+        "crs": bbox[-1],
+        "width": width,
+        "height": height,
+        "format": mime_type,
+        "transparent": True,
+    }
+
+    if styles is not None:
+        params["styles"] = styles
+
+    # create GetMap request
+    ogc_server_settings = OGC_Servers_Handler(settings.OGC_SERVER)["default"]
+
+    if ogc_server_location is not None:
+        thumbnail_url = ogc_server_location
+    else:
+        thumbnail_url = ogc_server_settings.LOCATION
+
+    wms_endpoint = ""
+    if thumbnail_url == ogc_server_settings.LOCATION:
+        # add access token to requests to Geoserver (logic based on the previous implementation)
+        username = ogc_server_settings.credentials.username
+        user = get_user_model().objects.filter(username=username).first()
+        if user:
+            access_token = get_or_create_token(user)
+            if access_token and not access_token.is_expired():
+                params["access_token"] = access_token.token
+
+        # add WMS endpoint to requests to Geoserver
+        wms_endpoint = getattr(ogc_server_settings, "WMS_ENDPOINT") or "ows"
+
+    thumbnail_url = thumbnail_url + f"{wms_endpoint}?" + "&".join(f"{key}={val}" for key, val in params.items())
+
+    return thumbnail_url
+
+
+def fetch_wms(url: str, max_retries: int = 3, retry_delay: int = 1):
+    """
+    Function fetching an image from OGC server. The request is performed based on the WMS URL.
+    In case access_token in not present in the URL , and Geoserver is used and the OGC backend, Basic Authentication
+    is used instead. If image retrieval fails, function retries to fetch the image max_retries times, waiting
+    retry_delay seconds between consecutive requests.
+
+    :param url: WMS URL of the image
+    :param max_retries: maximum number of retries before skipping retrieval
+    :param retry_delay: number of seconds waited between retries
+    :returns: retrieved image
+    """
+
+    # prepare authorization for WMS service
+    headers = {}
+    if "access_token" not in url:
```

Comment:
@bieganowski you are assuming that the background will be retrieved from the local Geoserver. This is wrong, since the background can be requested from external WMS services. 
And what you do here is dangerous we're leaking Authorization information to an external service in this case.

Reply:
there is an additional check for geoserver backend in the next line - if `access_token` is not in the URL and the OGC backend in not geoserver, nothing will be appended to the query

Reply:
check updated, to make sure only Geoserver queries will be updated with the Basic Auth. Anyway, I was wondering if that shouldn't be removed at all, since this part of code should never be reached (all geoserver queries should append `access_token` before, in the first place)... but this is directly inherited from the original implementation, so I decided not to remove it.

Reply:
@bieganowski AFAIK (and also @afabiani confirms this), this is the only case where a direct call to Geoserver on bhealf of a user is done. All the other cases go through the internal proxy, which automatically includes the access token. Howevere, since we want to obtain the maixmum throughput here, I would keep the injection of the access token local, since going through the proxy adds a significant delay and overload.